src/answerer.rs

```diff
@@ -0,0 +1,355 @@
+use std::{
+    sync::Arc,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use anyhow::Context;
+use carrot_commons::http_client::HttpClient;
+use diesel::{
+    r2d2::{ConnectionManager, Pool},
+    PgConnection,
+};
+use ethers::{
+    middleware::{Middleware, SignerMiddleware},
+    providers::{Http, Provider},
+    signers::LocalWallet,
+    types::Address,
+    utils,
+};
+use tokio::time::interval;
+use tracing::{info_span, Instrument};
+
+use crate::{
+    commons::{ChainConfig, ANSWERING_TASK_INTERVAL_SECONDS},
+    contracts::{defi_llama_oracle::DefiLlamaOracle, kpi_token::KPIToken},
+    db::models::{self, ActiveOracle},
+    specification,
+};
+
+pub async fn answer_active_oracles(
+    chain_id: u64,
+    chain_config: ChainConfig,
+    signer: Arc<SignerMiddleware<Provider<Http>, LocalWallet>>,
+    db_connection_pool: Pool<ConnectionManager<PgConnection>>,
+    defillama_http_client: Arc<HttpClient>,
+) -> anyhow::Result<()> {
+    let mut interval = interval(
+        chain_config
+            .answering_task_interval_seconds
+            .map(|seconds| Duration::from_secs(seconds))
+            .unwrap_or(ANSWERING_TASK_INTERVAL_SECONDS),
+    );
+
+    tracing::info!("answering active oracles");
+
+    loop {
+        interval.tick().await;
+
+        if let Err(error) = handle_active_oracles_answering(
+            chain_id,
+            signer.clone(),
+            db_connection_pool.clone(),
+            defillama_http_client.clone(),
+        )
+        .await
+        {
+            tracing::error!("error while handling active oracles answering: {:#}", error);
+        }
+    }
+}
+
+pub async fn handle_active_oracles_answering(
+    chain_id: u64,
+    signer: Arc<SignerMiddleware<Provider<Http>, LocalWallet>>,
+    db_connection_pool: Pool<ConnectionManager<PgConnection>>,
+    defillama_http_client: Arc<HttpClient>,
+) -> anyhow::Result<()> {
+    let mut db_connection = match db_connection_pool.get() {
+        Ok(connection) => connection,
+        Err(error) => {
+            tracing::error!("could not get new connection from pool: {:#}", error);
+            return Ok(());
+        }
+    };
+
+    let active_oracles =
+        match models::ActiveOracle::get_all_answerable_for_chain_id(&mut db_connection, chain_id) {
+            Ok(oracles) => oracles,
+            Err(error) => {
+                tracing::error!(
+                    "could not get currently active oracles in chain with id {}: {:#}",
+                    chain_id,
+                    error
+                );
+                return Ok(());
+            }
+        };
+    drop(db_connection);
+
+    let active_oracles_len = active_oracles.len();
```

Comment:
Should we just opt out here if `active_oracles_len == 0`?

Reply:
Addressed in df7929fc491aadbdce2a122c26a062dca3ac6bc5